File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-java/src/test/java/shopping/cart/CreateTableTestUtils.java
package shopping.cart;

import static java.util.concurrent.TimeUnit.SECONDS;

import akka.actor.typed.ActorSystem;
import akka.persistence.jdbc.testkit.javadsl.SchemaUtils;
import akka.projection.jdbc.javadsl.JdbcProjection;
import java.util.Objects;
import javax.persistence.EntityManager;
import org.slf4j.LoggerFactory;
import org.springframework.orm.jpa.EntityManagerFactoryUtils;
import org.springframework.orm.jpa.JpaTransactionManager;
import org.springframework.transaction.TransactionStatus;
import shopping.cart.repository.HibernateJdbcSession;

public class CreateTableTestUtils {

  public static void createTables(JpaTransactionManager transactionManager, ActorSystem<?> system)
      throws Exception {

    // create schemas
    // ok to block here, main test thread

    SchemaUtils.dropIfExists(system).toCompletableFuture().get(30, SECONDS);
    ;
    SchemaUtils.createIfNotExists(system).toCompletableFuture().get(30, SECONDS);

    JdbcProjection.dropOffsetTableIfExists(
            () -> new HibernateJdbcSession(transactionManager), system)
        .toCompletableFuture()
        .get(30, SECONDS);
    ;
    JdbcProjection.createOffsetTableIfNotExists(
            () -> new HibernateJdbcSession(transactionManager), system)
        .toCompletableFuture()
        .get(30, SECONDS);
    ;

    // NOTE: item_popularity is created by auto-created hibernate in tests
    deleteItemPopularity(transactionManager);

    LoggerFactory.getLogger(CreateTableTestUtils.class).info("Tables created");
  }

  private static void deleteItemPopularity(JpaTransactionManager transactionManager) {
    TransactionStatus transaction = transactionManager.getTransaction(null);
    EntityManager entityManager =
        EntityManagerFactoryUtils.getTransactionalEntityManager(
            Objects.requireNonNull(transactionManager.getEntityManagerFactory()));
    entityManager.createQuery("delete from ItemPopularity").executeUpdate();
    entityManager.close();
    transactionManager.commit(transaction);
  }
}

File: docs-source/docs/modules/how-to/examples/shopping-cart-service-cassandra-scala/src/test/scala/shopping/cart/CreateTableTestUtils.scala
package shopping.cart

import org.apache.pekko.actor.typed.ActorSystem
import org.apache.pekko.stream.connectors.cassandra.scaladsl.CassandraSessionRegistry
import shopping.cart.ItemPopularityRepositoryImpl.popularityTable

object CreateTableTestUtils {

  def createTables(system: ActorSystem[_]): Unit = {
    import org.slf4j.LoggerFactory
    import org.apache.pekko.projection.cassandra.scaladsl.CassandraProjection
    import scala.concurrent.Await
    import scala.concurrent.duration._

    // ok to block here, main thread
    Await.result(
      CassandraProjection.createOffsetTableIfNotExists()(system),
      30.seconds)

    // use same keyspace for the item_popularity table as the offset store
    val keyspace = system.settings.config
      .getString("pekko.projection.cassandra.offset-store.keyspace")
    val session =
      CassandraSessionRegistry(system).sessionFor("pekko.persistence.cassandra")

    Await.result(
      session.executeDDL(
        s"""CREATE TABLE IF NOT EXISTS $keyspace.$popularityTable (
      item_id text,
      count counter,
      PRIMARY KEY (item_id))
      """),
      30.seconds)

    LoggerFactory
      .getLogger("shopping.cart.CreateTableTestUtils")
      .info("Created keyspace [{}] and tables", keyspace)
  }
}
